pyartcd/pyartcd/pipelines/promote.py

```diff
@@ -274,6 +271,47 @@ def _reraise_if_not_permitted(self, err: VerificationError, code: str, permits:
         self._logger.warn("Issue %s is permitted with justification: %s", err, justification)
         return justification
 
+    def validate_upgrades(self, upgrades_str: str, assembly_type: assembly.AssemblyTypes, build_data_path: os.PathLike) -> List:
+        if upgrades_str is None and assembly_type != assembly.AssemblyTypes.CUSTOM:
+            raise ValueError(f"Group config for assembly {self.assembly} is missing the required `upgrades` field. If no upgrade edges are expected, please explicitly set the `upgrades` field to an empty string.")
+        previous_list = list(map(lambda s: s.strip(), upgrades_str.split(","))) if upgrades_str else []
+        # Ensure all versions in previous list are valid semvers.
+        if any(map(lambda version: not VersionInfo.isvalid(version), previous_list)):
+            raise ValueError("Previous list (`upgrades` field in group config) has an invalid semver.")
+
+        major, minor = util.isolate_major_minor_in_group(self.group)
+        if minor < 1:
+            return previous_list
+
+        prev_ver = f'{major}.{minor-1}'
+        show_spec = f"origin/openshift-{prev_ver}:releases.yml"
+        cmd = [
+            "git",
+            "show",
+            show_spec,
+        ]
+        proc = subprocess.run(cmd, capture_output=True, cwd=Path(build_data_path))
+        releases_config = yaml.safe_load(proc.stdout)
+
+        def looks_standard_upgrade_edge(config, x):
+            rel_type = config['releases'][x]['assembly'].get('type', '')
+            is_not_custom = rel_type != 'custom'
+            is_candidate = (rel_type == 'candidate') or (re.match(r'rc\.\d+', x) or re.match(r'fc\.\d+', x))
+            looks_standard = re.match(rf'{major}\.{minor-1}.\d+', x)
+            return is_not_custom and (is_candidate or looks_standard)
+
+        assembly_names = [x for x in releases_config['releases'].keys() if looks_standard_upgrade_edge(releases_config, x)]
+
+        def sort_semver(versions):
+            return sorted(versions, key=functools.cmp_to_key(semver.compare), reverse=True)
+        in_previous_list = sort_semver([x for x in assembly_names if x in previous_list])
+        not_in_previous_list = [x for x in assembly_names if x not in previous_list]
+        latest_prev = in_previous_list[0]
+        greater_than_latest_prev = [x for x in not_in_previous_list if semver.compare(x, latest_prev) == 1]  # if x > latest_prev
+        if greater_than_latest_prev:
+            raise ValueError(f"`upgrades` does not contain {greater_than_latest_prev} edge(s) defined in {show_spec}. These versions were found to be greater than the latest previous upgrade edge {latest_prev}")
```

Comment:
This should probably be  a `VerificationError`.

Reply:
I don't know what difference @vfreex intended between ValueError and VerificationError.. I can change them but don't know if I should look for something. @vfreex ?

Reply:
ValueError when value is missing
VerificationError when value is incorrect ?

Reply:
I don't have strong opinion on either option since we don't capture this exception for now.